[ONNX] Refactor perfect/nearest match criteria to allow optional inputs and disallow mismatch attributes

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106478

Fix #106057, except **Attribute dtype mismatch. E.g., alpha of aten.add.Tensor. -> Attribute: alpha INT vs FLOAT**.

Summarized the change
* Fill in defaults of attribute when `param_schema` is applied. This relaxes the matching on default attributes.
* Fill in None to optional input when `param_schema` is applied.
* Keep extra kwargs in attributes to make matching strictly.
* Allow input to be None when its dtype is `optiona[INPUT]`

The change comes with the guarantee from torchlib that attribute would never be None. For example, if `memory_format` is needed. The function should specify like this:
```python
@torch_op("aten::clone")
def aten_clone(
    self: TTensor, memory_format: str = ""  # pylint: disable=unused-argument
) -> TTensor:
    """clone(Tensor self, *, MemoryFormat? memory_format=None) -> Tensor"""

    return op.Identity(self)
```

Previous to this PR, opSchema matching didn't strictly guard the number of inputs/attributes to allow nearest match, which introduces the bug of dispatching `aten::div.Tensor` to `aten::div.default` disregarding the fact that `aten::div.Tensor` has an extra attibute `rounding_mode`. This PR fixes the issue with the new logic to perfect/nearest match. Particularly, strictly restrict the qualification of being nearest match candidate.

For each ONNX variants, we check these step by step:
1. Check if the function signature of inputs number is the same as the inputs.
2. Check if the function signature of attribute names is the same set of inputs.

If either of the above two criteria fails to meet, the ONNX variant is not a perfect match, nor a nearest match candidate (match_score=None)

3. Check if input dtype matches
4. Check if attribute dtype matches

If 3 and 4 are met, then this is a perfect match, otherwise, it's still considered a candidate of nearest match with a matching score.

## Case Study

### Optional Input
The dispatcher recognizes optional inputs. However, the input can't be ignored. None must be provided.
```python
# Perfect match is found
inputs = (Tensor([2, 3]), None)
aten_op(X: TTensor, Y: Optional[INT64]):
    ...
```
Real Case: aten::convolution
NOTE: There is/will not optional attribute in torchlib.

### Different attributes
If an attribute is provided with value, it's a must to match the attribute in function signature.
```python
# Not perfect match, nor nearest match
inputs = (Tensor([2, 3]),)
attributes = {"a":1, "b":2}
aten_op(X: TTensor, a: int):
    ...
```
Real Case: aten::div and aten::div.Tensor_mode

### Default attribute
Default attribute will fill in the value into inputs/attributes
```python
# Perfect match is found
inputs = (Tensor([2, 3]),)
attributes = {}
aten_op(X: TTensor, a: int = 3):
    ...
```
Real case: aten::clone

### Ignore attribute with None value
The attributes with None value will be ignored in matching.
```python
# Perfect match
inputs = (Tensor([2, 3]),)
attributes = {"a": None}
aten_op(X: TTensor):
    ...

# Not perfect match, but eligible for nearest match
inputs = (Tensor([2, 3]),)
attributes = {"a": None}
aten_op(X: TTensor, a: int = 3):
    ...
```
Real case: aten::div and aten::div.Tensor_mode

titaiwangms committed:
[ONNX] Refactor perfect/nearest match criteria to allow optional inputs and disallow mismatch attributes

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor perfect/nearest match criteria to allow optional inputs and disallow mismatch attributes"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor perfect/nearest match criteria to allow optional inputs and disallow mismatch attributes"




[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -359,7 +359,7 @@ def _convert_torch_args_to_onnxfunction_args(
             else:
                 tagged_kwargs[param.name] = kwargs[param.name]
-        elif param.default is not object():
+        elif param.default is not values._EmptyDefault:
```
Use `onnx.AttributeProto` to fetch Undefined instead.

titaiwangms commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -722,15 +759,28 @@ def _separate_input_attributes_from_arguments(
-                    onnx_attributes[param.name] = kwargs[param.name]
-            elif param.is_attribute and param.default is not object():
+                    onnx_attributes[param.name] = copy_kwargs[param.name]
+            elif param.is_attribute and param.default is not values._EmptyDefault:
```
Use onnx.AttributeProto to fetch Undefined instead.

titaiwangms committed:
Update on "[ONNX] Refactor perfect/nearest match criteria to allow optional inputs and disallow mismatch attributes"


Fix #106057, except **Attribute dtype mismatch. E.g., alpha of aten.add.Tensor. -> Attribute: alpha INT vs FLOAT**.

Summarized the change
* Fill in defaults of attribute when `param_schema` is applied. This relaxes the matching on default attributes.
* Fill in None to optional input when `param_schema` is applied.
* Keep extra kwargs in attributes to make matching strictly.
* Allow input to be None when its dtype is `optiona[INPUT]`

The change comes with the guarantee from torchlib that attribute would never be None. For example, if `memory_format` is needed. The function should specify like this:
```python
torch_op("aten::clone")
def aten_clone(
    self: TTensor, memory_format: str = ""  # pylint: disable=unused-argument
) -> TTensor:
    """clone(Tensor self, *, MemoryFormat? memory_format=None) -> Tensor"""

    return op.Identity(self)
```

Previous to this PR, opSchema matching didn't strictly guard the number of inputs/attributes to allow nearest match, which introduces the bug of dispatching `aten::div.Tensor` to `aten::div.default` disregarding the fact that `aten::div.Tensor` has an extra attibute `rounding_mode`. This PR fixes the issue with the new logic to perfect/nearest match. Particularly, strictly restrict the qualification of being nearest match candidate.

For each ONNX variants, we check these step by step:
1. Check if the function signature of inputs number is the same as the inputs.
2. Check if the function signature of attribute names is the same set of inputs.

If either of the above two criteria fails to meet, the ONNX variant is not a perfect match, nor a nearest match candidate (match_score=None)

3. Check if input dtype matches
4. Check if attribute dtype matches

If 3 and 4 are met, then this is a perfect match, otherwise, it's still considered a candidate of nearest match with a matching score.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor perfect/nearest match criteria to allow optional inputs and disallow mismatch attributes"


Fix #106057, except **Attribute dtype mismatch. E.g., alpha of aten.add.Tensor. -> Attribute: alpha INT vs FLOAT**.

Summarized the change
* Fill in defaults of attribute when `param_schema` is applied. This relaxes the matching on default attributes.
* Fill in None to optional input when `param_schema` is applied.
* Keep extra kwargs in attributes to make matching strictly.
* Allow input to be None when its dtype is `optiona[INPUT]`

The change comes with the guarantee from torchlib that attribute would never be None. For example, if `memory_format` is needed. The function should specify like this:
```python
torch_op("aten::clone")
def aten_clone(
    self: TTensor, memory_format: str = ""  # pylint: disable=unused-argument
) -> TTensor:
    """clone(Tensor self, *, MemoryFormat? memory_format=None) -> Tensor"""

    return op.Identity(self)
```

Previous to this PR, opSchema matching didn't strictly guard the number of inputs/attributes to allow nearest match, which introduces the bug of dispatching `aten::div.Tensor` to `aten::div.default` disregarding the fact that `aten::div.Tensor` has an extra attibute `rounding_mode`. This PR fixes the issue with the new logic to perfect/nearest match. Particularly, strictly restrict the qualification of being nearest match candidate.

For each ONNX variants, we check these step by step:
1. Check if the function signature of inputs number is the same as the inputs.
2. Check if the function signature of attribute names is the same set of inputs.

If either of the above two criteria fails to meet, the ONNX variant is not a perfect match, nor a nearest match candidate (match_score=None)

3. Check if input dtype matches
4. Check if attribute dtype matches

If 3 and 4 are met, then this is a perfect match, otherwise, it's still considered a candidate of nearest match with a matching score.

[ghstack-poisoned]

titaiwangms commented:
@BowenBao PTAL

BowenBao commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -352,6 +352,10 @@
     ),
+    xfail(
+        "nn.functional.adaptive_avg_pool1d",
+        reason=onnx_test_common.reason_onnx_script_does_not_support("aten::div.Tensor_mode"),
```
Plz create an issue under onnxscript for missing ops.. There is at least one model in benchmark needing it too. Thx!

titaiwangms replied:
https://github.com/microsoft/onnxscript/issues/980

abock replied:
This was merged, so remove the `xfail`?

titaiwangms replied:
Sure, I can update onnx-script version.

titaiwangms replied:
Tested. Found that we need onnx-script type promotion to support on this.

titaiwangms replied:
https://github.com/microsoft/onnxscript/issues/563

BowenBao commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -430,7 +446,7 @@ class _OnnxSchemaChecker:
     kwargs types.
 
-    There are three types of ONNX overloads in torchlib:
+    There are two types of ONNX overloads in torchlib:
```
```suggestion
    There are two types of ONNX variants in torchlib:
```
does this change make sense? If so we should adopt this convention to separate concept from aten overload.

BowenBao commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -310,16 +308,16 @@ def _convert_torch_args_to_onnxfunction_args(
 ) -> Tuple[List[Any], Dict[str, Any],]:
     """Convert Python args and kwargs to OnnxFunction acceptable with matching ONNX ParamSchema.
 
+    This is different from the para
```
incomplete

BowenBao commented on torch/onnx/_internal/fx/type_utils.py:
```diff
@@ -44,6 +44,15 @@ def from_sym_value_to_torch_dtype(sym_value: SYM_VALUE_TYPE) -> torch.dtype:
 
 
+def is_optional_onnx_dtype_str(onnx_type_str: str) -> bool:
+    optional_onnx_dtype_str = {
```
nit: I wonder if we should cache `optional_onnx_dtype_str ` or define it instead of recomputing every time.

titaiwangms replied:
We should. I missed this.

BowenBao commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -627,15 +663,18 @@ def _record_matching_score(
+        r"""Calculate the inputs matching score of the OpSchema requirements to find the nearest match.
+
+        Only the functions which have the same number of inputs and attributes as the
+        OpSchema are eligible to be a nearest match candidate. Thus, we don't need to
```
maybe check and raise otherwise in the beginning of this function

titaiwangms replied:
It's guaranteed to have the same length of inputs and attributes because `self._record_matching_score(function_inputs, function_attributes)` is only invoked when the check are done in `perfect_match` function (step 1).

BowenBao commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -590,6 +623,9 @@ def perfect_match_inputs(
                     f"Actual {type(attribute)} vs\n"
                     f"expected {self.attributes[attribute_name].type}\n"
                 )
+                # NOTE: This is still a candidate for nearest match, as it only mismatches on dtypes.
```
```suggestion
                # NOTE: This is still a candidate for nearest match, as it only mismatches on attribute dtypes.
```

BowenBao commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -569,18 +610,10 @@ def perfect_match_inputs(
                     f"Actual {torch_input_compatible_types} vs\n"
                     f"expected {allowed_types}\n"
                 )
+                # NOTE: This is still a candidate for nearest match, as it only mismatches on dtypes.
```
```suggestion
                # NOTE: This is still a candidate for nearest match, as it only mismatches on input dtypes.
```

BowenBao commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -569,18 +610,10 @@ def perfect_match_inputs(
+                # NOTE: This is still a candidate for nearest match, as it only mismatches on dtypes.
+                self._record_matching_score(function_inputs, function_attributes)
+                diagnostic.with_additional_message(f"match score: {self.match_score}\n")
                 return False
```
we are short circuiting the diagnostic message for attr dtype mismatches, if there is input dtype mismatch...

titaiwangms replied:
I use a flag `is_perfect_match` to postpone the return to let diagnostics catch all it needs.

titaiwangms committed:
Update on "[ONNX] Refactor perfect/nearest match criteria to allow optional inputs and disallow mismatch attributes"


Fix #106057, except **Attribute dtype mismatch. E.g., alpha of aten.add.Tensor. -> Attribute: alpha INT vs FLOAT**.

Summarized the change
* Fill in defaults of attribute when `param_schema` is applied. This relaxes the matching on default attributes.
* Fill in None to optional input when `param_schema` is applied.
* Keep extra kwargs in attributes to make matching strictly.
* Allow input to be None when its dtype is `optiona[INPUT]`

The change comes with the guarantee from torchlib that attribute would never be None. For example, if `memory_format` is needed. The function should specify like this:
```python
torch_op("aten::clone")
def aten_clone(
    self: TTensor, memory_format: str = ""  # pylint: disable=unused-argument
) -> TTensor:
    """clone(Tensor self, *, MemoryFormat? memory_format=None) -> Tensor"""

    return op.Identity(self)
```

Previous to this PR, opSchema matching didn't strictly guard the number of inputs/attributes to allow nearest match, which introduces the bug of dispatching `aten::div.Tensor` to `aten::div.default` disregarding the fact that `aten::div.Tensor` has an extra attibute `rounding_mode`. This PR fixes the issue with the new logic to perfect/nearest match. Particularly, strictly restrict the qualification of being nearest match candidate.

For each ONNX variants, we check these step by step:
1. Check if the function signature of inputs number is the same as the inputs.
2. Check if the function signature of attribute names is the same set of inputs.

If either of the above two criteria fails to meet, the ONNX variant is not a perfect match, nor a nearest match candidate (match_score=None)

3. Check if input dtype matches
4. Check if attribute dtype matches

If 3 and 4 are met, then this is a perfect match, otherwise, it's still considered a candidate of nearest match with a matching score.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor perfect/nearest match criteria to allow optional inputs and disallow mismatch attributes"


Fix #106057, except **Attribute dtype mismatch. E.g., alpha of aten.add.Tensor. -> Attribute: alpha INT vs FLOAT**.

Summarized the change
* Fill in defaults of attribute when `param_schema` is applied. This relaxes the matching on default attributes.
* Fill in None to optional input when `param_schema` is applied.
* Keep extra kwargs in attributes to make matching strictly.
* Allow input to be None when its dtype is `optiona[INPUT]`

The change comes with the guarantee from torchlib that attribute would never be None. For example, if `memory_format` is needed. The function should specify like this:
```python
torch_op("aten::clone")
def aten_clone(
    self: TTensor, memory_format: str = ""  # pylint: disable=unused-argument
) -> TTensor:
    """clone(Tensor self, *, MemoryFormat? memory_format=None) -> Tensor"""

    return op.Identity(self)
```

Previous to this PR, opSchema matching didn't strictly guard the number of inputs/attributes to allow nearest match, which introduces the bug of dispatching `aten::div.Tensor` to `aten::div.default` disregarding the fact that `aten::div.Tensor` has an extra attibute `rounding_mode`. This PR fixes the issue with the new logic to perfect/nearest match. Particularly, strictly restrict the qualification of being nearest match candidate.

For each ONNX variants, we check these step by step:
1. Check if the function signature of inputs number is the same as the inputs.
2. Check if the function signature of attribute names is the same set of inputs.

If either of the above two criteria fails to meet, the ONNX variant is not a perfect match, nor a nearest match candidate (match_score=None)

3. Check if input dtype matches
4. Check if attribute dtype matches

If 3 and 4 are met, then this is a perfect match, otherwise, it's still considered a candidate of nearest match with a matching score.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor perfect/nearest match criteria to allow optional inputs and disallow mismatch attributes"


Fix #106057, except **Attribute dtype mismatch. E.g., alpha of aten.add.Tensor. -> Attribute: alpha INT vs FLOAT**.

Summarized the change
* Fill in defaults of attribute when `param_schema` is applied. This relaxes the matching on default attributes.
* Fill in None to optional input when `param_schema` is applied.
* Keep extra kwargs in attributes to make matching strictly.
* Allow input to be None when its dtype is `optiona[INPUT]`

The change comes with the guarantee from torchlib that attribute would never be None. For example, if `memory_format` is needed. The function should specify like this:
```python
torch_op("aten::clone")
def aten_clone(
    self: TTensor, memory_format: str = ""  # pylint: disable=unused-argument
) -> TTensor:
    """clone(Tensor self, *, MemoryFormat? memory_format=None) -> Tensor"""

    return op.Identity(self)
```

Previous to this PR, opSchema matching didn't strictly guard the number of inputs/attributes to allow nearest match, which introduces the bug of dispatching `aten::div.Tensor` to `aten::div.default` disregarding the fact that `aten::div.Tensor` has an extra attibute `rounding_mode`. This PR fixes the issue with the new logic to perfect/nearest match. Particularly, strictly restrict the qualification of being nearest match candidate.

For each ONNX variants, we check these step by step:
1. Check if the function signature of inputs number is the same as the inputs.
2. Check if the function signature of attribute names is the same set of inputs.

If either of the above two criteria fails to meet, the ONNX variant is not a perfect match, nor a nearest match candidate (match_score=None)

3. Check if input dtype matches
4. Check if attribute dtype matches

If 3 and 4 are met, then this is a perfect match, otherwise, it's still considered a candidate of nearest match with a matching score.

## Case Study

### Optional Input
The dispatcher recognizes optional inputs. However, the input can't be ignored. None must be provided.
```python
# Perfect match is found
inputs = (Tensor([2, 3]), None)
aten_op(X: TTensor, Y: Optional[INT64]):
    ...
```
Real Case: aten::convolution
NOTE: There is/will not optional attribute in torchlib.

### Different attributes
If an attribute is provided with value, it's a must to match the attribute in function signature.
```python
# Not perfect match, nor nearest match
inputs = (Tensor([2, 3]),)
attributes = {"a":1, "b":2}
aten_op(X: TTensor, a: int):
    ...
```
Real Case: aten::div and aten::div.Tensor_mode

### Default attribute
Default attribute will fill in the value into inputs/attributes
```python
# Perfect match is found
inputs = (Tensor([2, 3]),)
attributes = {}
aten_op(X: TTensor, a: int = 3):
    ...
```
Real case: aten::clone

### Ignore attribute with None value
The attributes with None value will be ignored in matching.
```python
# Perfect match
inputs = (Tensor([2, 3]),)
attributes = {"a": None}
aten_op(X: TTensor):
    ...

# Not perfect match, but eligible for nearest match
inputs = (Tensor([2, 3]),)
attributes = {"a": None}
aten_op(X: TTensor, a: int = 3):
    ...
```
Real case: aten::div and aten::div.Tensor_mode


[ghstack-poisoned]

titaiwangms commented:
@BowenBao added case study which should be enough to explain the situations that this PR improves. Essentially, most of the changes are to align with the expectations of attribute from torchlib.

BowenBao commented:
Thanks this is great. Could you put the case studies into comment and create a [NOTE: title] for it? There are also some comments I want to make.

thiagocrepaldi reviewed (approved):
LGTM. Just make sure Bowen's comments are addressed before merging :)

titaiwangms committed:
Update on "[ONNX] Refactor perfect/nearest match criteria to allow optional inputs and disallow mismatch attributes"


Fix #106057, except **Attribute dtype mismatch. E.g., alpha of aten.add.Tensor. -> Attribute: alpha INT vs FLOAT**.

Summarized the change
* Fill in defaults of attribute when `param_schema` is applied. This relaxes the matching on default attributes.
* Fill in None to optional input when `param_schema` is applied.
* Keep extra kwargs in attributes to make matching strictly.
* Allow input to be None when its dtype is `optiona[INPUT]`

The change comes with the guarantee from torchlib that attribute would never be None. For example, if `memory_format` is needed. The function should specify like this:
```python
torch_op("aten::clone")
def aten_clone(
    self: TTensor, memory_format: str = ""  # pylint: disable=unused-argument
) -> TTensor:
    """clone(Tensor self, *, MemoryFormat? memory_format=None) -> Tensor"""

    return op.Identity(self)
```

Previous to this PR, opSchema matching didn't strictly guard the number of inputs/attributes to allow nearest match, which introduces the bug of dispatching `aten::div.Tensor` to `aten::div.default` disregarding the fact that `aten::div.Tensor` has an extra attibute `rounding_mode`. This PR fixes the issue with the new logic to perfect/nearest match. Particularly, strictly restrict the qualification of being nearest match candidate.

For each ONNX variants, we check these step by step:
1. Check if the function signature of inputs number is the same as the inputs.
2. Check if the function signature of attribute names is the same set of inputs.

If either of the above two criteria fails to meet, the ONNX variant is not a perfect match, nor a nearest match candidate (match_score=None)

3. Check if input dtype matches
4. Check if attribute dtype matches

If 3 and 4 are met, then this is a perfect match, otherwise, it's still considered a candidate of nearest match with a matching score.

## Case Study

### Optional Input
The dispatcher recognizes optional inputs. However, the input can't be ignored. None must be provided.
```python
# Perfect match is found
inputs = (Tensor([2, 3]), None)
aten_op(X: TTensor, Y: Optional[INT64]):
    ...
```
Real Case: aten::convolution
NOTE: There is/will not optional attribute in torchlib.

### Different attributes
If an attribute is provided with value, it's a must to match the attribute in function signature.
```python
# Not perfect match, nor nearest match
inputs = (Tensor([2, 3]),)
attributes = {"a":1, "b":2}
aten_op(X: TTensor, a: int):
    ...
```
Real Case: aten::div and aten::div.Tensor_mode

### Default attribute
Default attribute will fill in the value into inputs/attributes
```python
# Perfect match is found
inputs = (Tensor([2, 3]),)
attributes = {}
aten_op(X: TTensor, a: int = 3):
    ...
```
Real case: aten::clone

### Ignore attribute with None value
The attributes with None value will be ignored in matching.
```python
# Perfect match
inputs = (Tensor([2, 3]),)
attributes = {"a": None}
aten_op(X: TTensor):
    ...

# Not perfect match, but eligible for nearest match
inputs = (Tensor([2, 3]),)
attributes = {"a": None}
aten_op(X: TTensor, a: int = 3):
    ...
```
Real case: aten::div and aten::div.Tensor_mode


[ghstack-poisoned]

titaiwangms commented:
> Thanks this is great. Could you put the case studies into comment and create a [NOTE: title] for it? There are also some comments I want to make.

Done. I extend/modify the description on _OpschemaWrapper

BowenBao commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -431,38 +447,84 @@ class _OnnxSchemaChecker:
+        Result: Perfect match.
+
+    2. [NOTE: Optional input]: The dispatcher recognizes optional inputs. However,
+        the input can't be ignored. None must be provided.
```
nit: to check my understanding. Is this due to onnxscript limitation that None cannot be set as default value for `Optional[xxx]` typed argument? Such that `inputs = (Tensor([2, 3], )` will not be matched, but otherwise it should? If there is an issue to link here that will be great ...

I typed above but then I found convolution aten signature is `- func: convolution(Tensor input, Tensor weight, Tensor? bias, ...`. So the optional input by design does not have default value. And in that case, the type is only meaning `None | INT64` so it is syntactically incorrect to omit it.

titaiwangms replied:
I think PyTorch doesn't omit any inputs, because if they do, we should see a lot of breakings now in terms of the `param_schema` separating. Not sure if there will be a change on this, but currently, we don't fill in default value to input, nor omit None.

BowenBao commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -431,38 +447,84 @@ class _OnnxSchemaChecker:
+        Result: Perfect match.
+        Real example: `aten::clone`
+
+    5. [NOTE: Ignore attribute with None value]: The attributes with None value
```
I guess this is due to the ONNXScript limitation I typed above? Cannot define optional attribute?

titaiwangms replied:
I think it's something about AttributeProto, but yes, we don't have Optional attribute at the time, and this assumption is acknowledged by OpSchemaWrapper.

Semantically, besides dtype, another goal of creating ONNX variants is to overload the cases with different attributes, because they are usaully composed by different set of ops. Thus, attribute with None value simply means the function doesn't need this attribute.

cc @justinchuby

justinchuby replied:
I will need to double check with Rama on optional attributes. IIRC we either provide them in the graph or not provide them in the graph. There is not a "None" as the attribute value. In the torchscript builder, we omit None attributes whenever we see them.

BowenBao commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -540,30 +612,54 @@ def perfect_match_inputs(
-
+        # NOTE: 1. Check if the input number and attribute names match the
+        # OpSchema. If it's not, we know the function is not eligible to be a perfect
+        # match, and even a nearest match.
```
```suggestion
        # match, nor a nearest match.
```

BowenBao reviewed (approved):
LG w/ comments.

titaiwangms committed:
Update on "[ONNX] Refactor perfect/nearest match criteria to allow optional inputs and disallow mismatch attributes"


Fix #106057, except **Attribute dtype mismatch. E.g., alpha of aten.add.Tensor. -> Attribute: alpha INT vs FLOAT**.

Summarized the change
* Fill in defaults of attribute when `param_schema` is applied. This relaxes the matching on default attributes.
* Fill in None to optional input when `param_schema` is applied.
* Keep extra kwargs in attributes to make matching strictly.
* Allow input to be None when its dtype is `optiona[INPUT]`

The change comes with the guarantee from torchlib that attribute would never be None. For example, if `memory_format` is needed. The function should specify like this:
```python
torch_op("aten::clone")
def aten_clone(
    self: TTensor, memory_format: str = ""  # pylint: disable=unused-argument
) -> TTensor:
    """clone(Tensor self, *, MemoryFormat? memory_format=None) -> Tensor"""

    return op.Identity(self)
```

Previous to this PR, opSchema matching didn't strictly guard the number of inputs/attributes to allow nearest match, which introduces the bug of dispatching `aten::div.Tensor` to `aten::div.default` disregarding the fact that `aten::div.Tensor` has an extra attibute `rounding_mode`. This PR fixes the issue with the new logic to perfect/nearest match. Particularly, strictly restrict the qualification of being nearest match candidate.

For each ONNX variants, we check these step by step:
1. Check if the function signature of inputs number is the same as the inputs.
2. Check if the function signature of attribute names is the same set of inputs.

If either of the above two criteria fails to meet, the ONNX variant is not a perfect match, nor a nearest match candidate (match_score=None)

3. Check if input dtype matches
4. Check if attribute dtype matches

If 3 and 4 are met, then this is a perfect match, otherwise, it's still considered a candidate of nearest match with a matching score.

## Case Study

### Optional Input
The dispatcher recognizes optional inputs. However, the input can't be ignored. None must be provided.
```python
# Perfect match is found
inputs = (Tensor([2, 3]), None)
aten_op(X: TTensor, Y: Optional[INT64]):
    ...
```
Real Case: aten::convolution
NOTE: There is/will not optional attribute in torchlib.

### Different attributes
If an attribute is provided with value, it's a must to match the attribute in function signature.
```python
# Not perfect match, nor nearest match
inputs = (Tensor([2, 3]),)
attributes = {"a":1, "b":2}
aten_op(X: TTensor, a: int):
    ...
```
Real Case: aten::div and aten::div.Tensor_mode

### Default attribute
Default attribute will fill in the value into inputs/attributes
```python
# Perfect match is found
inputs = (Tensor([2, 3]),)
attributes = {}
aten_op(X: TTensor, a: int = 3):
    ...
```
Real case: aten::clone

### Ignore attribute with None value
The attributes with None value will be ignored in matching.
```python
# Perfect match
inputs = (Tensor([2, 3]),)
attributes = {"a": None}
aten_op(X: TTensor):
    ...

# Not perfect match, but eligible for nearest match
inputs = (Tensor([2, 3]),)
attributes = {"a": None}
aten_op(X: TTensor, a: int = 3):
    ...
```
Real case: aten::div and aten::div.Tensor_mode


[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge
